lib/python/numpy.h

```diff
@@ -85,54 +76,144 @@ auto cast_to_array_like(const py::object &obj, const units::Unit unit) {
   }
 }
 
-template <bool convert, class T, class View>
-void copy_flattened_0d(const py::array_t<T> &data, View &&view) {
-  auto r = data.unchecked();
-  auto it = view.begin();
-  copy_element<convert>(r(), *it);
+namespace scipp::detail {
+namespace {
+template <class T> bool is_c_contiguous(const py::array_t<T> &array) {
+  Py_buffer buffer;
+  if (PyObject_GetBuffer(array.ptr(), &buffer, PyBUF_C_CONTIGUOUS) != 0) {
+    PyErr_Clear();
+    return false;
+  }
+  PyBuffer_Release(&buffer);
+  return true;
 }
 
-template <bool convert, class T, class View>
-void copy_flattened_1d(const py::array_t<T> &data, View &&view) {
-  auto r = data.unchecked();
-  auto it = view.begin();
-  for (scipp::index i = 0; i < r.shape(0); ++i, ++it) {
-    copy_element<convert>(r(i), *it);
+template <bool convert, class Source, class Destination>
+void copy_element(const Source &src, Destination &&dst) {
+  if constexpr (convert) {
+    dst = std::remove_reference_t<Destination>{src};
+  } else {
+    std::forward<Destination>(dst) = src;
   }
 }
 
-template <bool convert, class T, class View>
-void copy_flattened_2d(const py::array_t<T> &data, View &&view) {
-  auto r = data.unchecked();
-  const auto begin = view.begin();
+template <bool convert, class T, class Dst>
+void copy_flattened_0d(const py::array_t<T> &src_array, Dst &dst) {
+  const auto src = src_array.template unchecked<0>();
+  auto it = dst.begin();
+  copy_element<convert>(src(), *it);
+}
+
+template <bool convert, class T, class Dst>
+void copy_flattened_1d(const py::array_t<T> &src_array, Dst &dst) {
+  const auto src = src_array.template unchecked<1>();
+  const auto begin = dst.begin();
+  core::parallel::parallel_for(
+      core::parallel::blocked_range(0, src.shape(0), 10000),
+      [&](const auto &range) {
+        auto it = begin + range.begin();
+        for (scipp::index i = range.begin(); i < range.end(); ++i, ++it) {
+          copy_element<convert>(src(i), *it);
+        }
+      });
+}
+
+template <bool convert, class T, class Dst>
+void copy_flattened_2d(const py::array_t<T> &src_array, Dst &dst) {
+  const auto src = src_array.template unchecked<2>();
+  const auto begin = dst.begin();
+  core::parallel::parallel_for(
+      core::parallel::blocked_range(0, src.shape(0), 10000),
```

Comment:
Isn't this disabling multi-threading if `shape[0] < 10000`? Shouldn't we use the default setup (which, iirc, is `shape[0] / 24`, i.e., threaded for `shape[0]>1`? Or is this harmful in more important cases?

Same comment applies to 3d+ cases.